File: frontend/src/modeling/lib/find-exterior-faces.ts
import type { Vector3 } from "@/state/types";
import { faces } from "./voxel-constants";
import { getTextureCoordinates } from "./texture-coords";
import { MeshArrays } from "./mesh-arrays";
import {
  calculateAmbientOcclusion,
  precomputeAoOffsets,
  OCCLUSION_LEVELS,
} from "./ambient-occlusion";

export const DISABLE_GREEDY_MESHING = false;



export class ExteriorFacesFinder {
  private mask: Int16Array;
  private processed: Uint8Array;
  private aoMask: Uint8Array;
  private isSelectedMask: Uint8Array;
  private maskSize: number;
  private maxDim: number;

  constructor(maxDimension: number) {
    this.maskSize = maxDimension ** 2;
    this.maxDim = maxDimension;
    this.mask = new Int16Array(this.maskSize);
    this.processed = new Uint8Array(this.maskSize);
    this.aoMask = new Uint8Array(this.maskSize);
    this.isSelectedMask = new Uint8Array(this.maskSize);
  }

  public findExteriorFaces(
    voxelData: Uint8Array,
    textureWidth: number,
    blockAtlasMapping: number[],
    dimensions: Vector3,
    meshArrays: MeshArrays,
    selectionBuffer: Uint8Array,
    selectionWorldDims: Vector3,
    chunkOffset: Vector3,
    selectionEmpty: boolean
  ): void {
    meshArrays.reset();

    const maxDimension = Math.max(dimensions.x, dimensions.y, dimensions.z);
    const currentMaskSize = maxDimension ** 2;

    if (currentMaskSize > this.maskSize) {
      this.maskSize = currentMaskSize;
      this.maxDim = maxDimension;
      this.mask = new Int16Array(currentMaskSize);
      this.processed = new Uint8Array(currentMaskSize);
      this.aoMask = new Uint8Array(currentMaskSize);
      this.isSelectedMask = new Uint8Array(currentMaskSize);
    }

    const dimX = dimensions.x;
    const dimY = dimensions.y;
    const dimZ = dimensions.z;
    const strideX = dimY * dimZ;
    const maxDim = this.maxDim;
    const selWorldYZ = selectionWorldDims.y * selectionWorldDims.z;
    const selWorldZ = selectionWorldDims.z;
    const offX = chunkOffset.x;
    const offY = chunkOffset.y;
    const offZ = chunkOffset.z;

    for (let axis = 0; axis < 3; axis++) {
      const u = (axis + 1) % 3;
      const v = (axis + 2) % 3;

      const axisSize = axis === 0 ? dimX : axis === 1 ? dimY : dimZ;
      const uSize = u === 0 ? dimX : u === 1 ? dimY : dimZ;
      const vSize = v === 0 ? dimX : v === 1 ? dimY : dimZ;

      for (let dir = -1; dir <= 1; dir += 2) {
        const faceDir = axis * 2 + (dir > 0 ? 0 : 1);
        const dx = axis === 0 ? dir : 0;
        const dy = axis === 1 ? dir : 0;
        const dz = axis === 2 ? dir : 0;
        const neighborMax = axis === 0 ? dimX : axis === 1 ? dimY : dimZ;
        const aoOffsets = precomputeAoOffsets(faceDir, strideX, dimZ);
        const normalFlatOffset = dx * strideX + dy * dimZ + dz;
        
        const xIsDepth = axis === 0;
        const yIsDepth = axis === 1;
        const zIsDepth = axis === 2;
        const xIsUAxis = !xIsDepth && u === 0;
        const yIsUAxis = !yIsDepth && u === 1;
        const zIsUAxis = !zIsDepth && u === 2;

        for (let d = 0; d < axisSize; d++) {
          for (let iv = 0; iv < vSize; iv++) {
            const rowOffset = iv * maxDim;
            this.mask.fill(-1, rowOffset, rowOffset + uSize);
            this.aoMask.fill(0, rowOffset, rowOffset + uSize);
            this.isSelectedMask.fill(0, rowOffset, rowOffset + uSize);
          }

          let hasFaces = false;

          for (let iu = 0; iu < uSize; iu++) {
            for (let iv = 0; iv < vSize; iv++) {
              const x = xIsDepth ? d : xIsUAxis ? iu : iv;
              const y = yIsDepth ? d : yIsUAxis ? iu : iv;
              const z = zIsDepth ? d : zIsUAxis ? iu : iv;

              const blockIdx = x * strideX + y * dimZ + z;
              const blockValue = voxelData[blockIdx];
              const blockType = blockValue & 0x7F;
              const blockVisible = blockType !== 0;
              const blockIsSelected = selectionEmpty ? false : selectionBuffer[(offX + x) * selWorldYZ + (offY + y) * selWorldZ + (offZ + z)] !== 0;

              if (!blockVisible && !blockIsSelected) {
                continue;
              }

              const nx = x + dx;
              const ny = y + dy;
              const nz = z + dz;
              const neighborIdx = blockIdx + normalFlatOffset;

              const maskIdx = iv * maxDim + iu;

              if (blockIsSelected && !blockVisible) {
                const neighborCoord = xIsDepth ? nx : yIsDepth ? ny : nz;
                const neighborInBounds = dir > 0 ? neighborCoord < neighborMax : neighborCoord >= 0;
                const neighborIsSelected = neighborInBounds && selectionBuffer[(offX + nx) * selWorldYZ + (offY + ny) * selWorldZ + (offZ + nz)] !== 0;

                if (!neighborIsSelected) {
                  const selectionBlockType = selectionBuffer[(offX + x) * selWorldYZ + (offY + y) * selWorldZ + (offZ + z)] & 0x7F;
                  const textureIndex =
                    blockAtlasMapping[Math.max(selectionBlockType, 1) - 1];
                  
                  this.aoMask[maskIdx] = calculateAmbientOcclusion(
                    nx, ny, nz, voxelData, dimX, dimY, dimZ, neighborIdx, aoOffsets
                  );
                  
                  this.mask[maskIdx] = textureIndex;
                  this.isSelectedMask[maskIdx] = 1;
                  hasFaces = true;
                }
              } else if (blockVisible) {
                const neighborCoord = xIsDepth ? nx : yIsDepth ? ny : nz;
                const neighborInBounds = dir > 0 ? neighborCoord < neighborMax : neighborCoord >= 0;
                const neighborVisible = neighborInBounds && (voxelData[neighborIdx] & 0x7F) !== 0;

                if (!neighborVisible) {
                  const textureIndex = blockAtlasMapping[blockType - 1];

                  this.aoMask[maskIdx] = calculateAmbientOcclusion(
                    nx, ny, nz, voxelData, dimX, dimY, dimZ, neighborIdx, aoOffsets
                  );

                  this.mask[maskIdx] = textureIndex;
                  if (blockIsSelected) {
                    this.isSelectedMask[maskIdx] = 1;
                  }
                  hasFaces = true;
                }
              }
            }
          }

          if (hasFaces) {
            this.generateGreedyMesh(
              this.mask,
              this.aoMask,
              this.isSelectedMask,
              this.processed,
              uSize,
              vSize,
              d,
              axis,
              u,
              v,
              dir,
              faceDir,
              textureWidth,
              meshArrays
            );
          }
        }
      }
    }
  }

  private generateGreedyMesh(
    mask: Int16Array,
    aoMask: Uint8Array,
    isSelectedMask: Uint8Array,
    processed: Uint8Array,
    width: number,
    height: number,
    depth: number,
    axis: number,
    u: number,
    v: number,
    dir: number,
    faceDir: number,
    textureWidth: number,
    meshArrays: MeshArrays
  ): void {
    const stride = this.maxDim;
    const faceData = faces[faceDir];
    const normal = faceData.normal;
    const faceOffset = dir > 0 ? 1 : 0;
    
    for (let iv = 0; iv < height; iv++) {
      const rowOffset = iv * stride;
      processed.fill(0, rowOffset, rowOffset + width);
    }

    for (let j = 0; j < height; j++) {
      const jOffset = j * stride;
      for (let i = 0; i < width; ) {
        const ji = jOffset + i;
        if (processed[ji] || mask[ji] < 0) {
          i++;
          continue;
        }

        const textureIndex = mask[ji];
        const isSelected = isSelectedMask[ji];
        const aoVal = aoMask[ji];
        let quadWidth = 1;
        if (!DISABLE_GREEDY_MESHING) {
          while (i + quadWidth < width) {
            const idx = jOffset + i + quadWidth;
            if (
              processed[idx] ||
              mask[idx] !== textureIndex ||
              aoMask[idx] !== aoVal ||
              isSelectedMask[idx] !== isSelected
            )
              break;
            quadWidth++;
          }
        }

        let quadHeight = 1;
        if (!DISABLE_GREEDY_MESHING) {
          outer: while (j + quadHeight < height) {
            const rowOff = (j + quadHeight) * stride;
            for (let w = 0; w < quadWidth; w++) {
              const idx = rowOff + i + w;
              if (
                processed[idx] ||
                mask[idx] !== textureIndex ||
                aoMask[idx] !== aoVal ||
                isSelectedMask[idx] !== isSelected
              )
                break outer;
            }
            quadHeight++;
          }
        }

        const endI = i + quadWidth;
        const endJ = j + quadHeight;
        for (let jj = j; jj < endJ; jj++) {
          const rowOff = jj * stride;
          for (let ii = i; ii < endI; ii++) {
            processed[rowOff + ii] = 1;
          }
        }

        const x = axis === 0 ? depth : u === 0 ? i : v === 0 ? j : 0;
        const y = axis === 1 ? depth : u === 1 ? i : v === 1 ? j : 0;
        const z = axis === 2 ? depth : u === 2 ? i : v === 2 ? j : 0;

        const textureCoords = getTextureCoordinates(textureIndex, textureWidth);

        const startVertexIndex = meshArrays.vertexCount;

        for (let vi = 0; vi < 4; vi++) {
          let actualVi = vi;
          if (dir < 0 && (vi === 1 || vi === 3)) {
            actualVi = vi === 1 ? 3 : 1;
          }

          let vx: number, vy: number, vz: number;

          if (axis === 0) {
            vx = x + faceOffset;
            if (actualVi === 0) {
              vy = y;
              vz = z;
            } else if (actualVi === 1) {
              vy = y + (u === 1 ? quadWidth : 0);
              vz = z + (u === 2 ? quadWidth : 0);
            } else if (actualVi === 2) {
              vy = y + (u === 1 ? quadWidth : v === 1 ? quadHeight : 0);
              vz = z + (u === 2 ? quadWidth : v === 2 ? quadHeight : 0);
            } else {
              vy = y + (v === 1 ? quadHeight : 0);
              vz = z + (v === 2 ? quadHeight : 0);
            }
          } else if (axis === 1) {
            vy = y + faceOffset;
            if (actualVi === 0) {
              vx = x;
              vz = z;
            } else if (actualVi === 1) {
              vx = x + (u === 0 ? quadWidth : 0);
              vz = z + (u === 2 ? quadWidth : 0);
            } else if (actualVi === 2) {
              vx = x + (u === 0 ? quadWidth : v === 0 ? quadHeight : 0);
              vz = z + (u === 2 ? quadWidth : v === 2 ? quadHeight : 0);
            } else {
              vx = x + (v === 0 ? quadHeight : 0);
              vz = z + (v === 2 ? quadHeight : 0);
            }
          } else {
            vz = z + faceOffset;
            if (actualVi === 0) {
              vx = x;
              vy = y;
            } else if (actualVi === 1) {
              vx = x + (u === 0 ? quadWidth : 0);
              vy = y + (u === 1 ? quadWidth : 0);
            } else if (actualVi === 2) {
              vx = x + (u === 0 ? quadWidth : v === 0 ? quadHeight : 0);
              vy = y + (u === 1 ? quadWidth : v === 1 ? quadHeight : 0);
            } else {
              vx = x + (v === 0 ? quadHeight : 0);
              vy = y + (v === 1 ? quadHeight : 0);
            }
          }

          meshArrays.pushVertex(vx, vy, vz);
          meshArrays.pushNormal(normal[0], normal[1], normal[2]);
          meshArrays.pushUV(textureCoords[vi * 2], textureCoords[vi * 2 + 1]);

          const packedAO = aoVal;

          let aoCornerIndex: number;
          if (faceDir === 1 || faceDir === 2 || faceDir === 5) {
            aoCornerIndex = vi === 1 ? 3 : vi === 3 ? 1 : vi;
          } else {
            aoCornerIndex = vi;
          }

          const occlusionCount = (packedAO >> (aoCornerIndex * 2)) & 0x03;
          const aoFactor = OCCLUSION_LEVELS[occlusionCount];
          meshArrays.pushAO(aoFactor);
          meshArrays.pushIsSelected(isSelected);
          meshArrays.incrementVertex();
        }

        meshArrays.pushIndex(startVertexIndex);
        meshArrays.pushIndex(startVertexIndex + 1);
        meshArrays.pushIndex(startVertexIndex + 2);
        meshArrays.pushIndex(startVertexIndex);
        meshArrays.pushIndex(startVertexIndex + 2);
        meshArrays.pushIndex(startVertexIndex + 3);

        i += quadWidth;
      }
    }
  }
}

File: wasm/src/find_exterior_faces.rs
use crate::ambient_occlusion::{calculate_ambient_occlusion, OCCLUSION_LEVELS};
use crate::mesh_arrays::MeshArrays;
use crate::texture_coords::get_texture_coordinates;
use crate::voxel_constants::FACES;

const PACKED_PRESENT_BIT: i32 = 1 << 19;

pub struct ExteriorFacesFinder {
    packed_mask: Vec<i32>,
    processed: Vec<u8>,
    mask_size: usize,
    max_dim: usize,
}

impl ExteriorFacesFinder {
    pub fn new(max_dimension: usize) -> Self {
        let mask_size = max_dimension * max_dimension;
        Self {
            packed_mask: vec![0; mask_size],
            processed: vec![0; mask_size],
            mask_size,
            max_dim: max_dimension,
        }
    }

    pub fn find_exterior_faces(
        &mut self,
        voxel_data: &[u8],
        texture_width: i32,
        block_atlas_mapping: &[i32],
        dim_x: usize,
        dim_y: usize,
        dim_z: usize,
        mesh_arrays: &mut MeshArrays,
        selection_data: &[u8],
        selection_dim_x: usize,
        selection_dim_y: usize,
        selection_dim_z: usize,
        selection_empty: bool,
    ) {
        mesh_arrays.reset();

        let max_dimension = dim_x.max(dim_y).max(dim_z);
        let current_mask_size = max_dimension * max_dimension;

        if current_mask_size > self.mask_size {
            self.mask_size = current_mask_size;
            self.max_dim = max_dimension;
            self.packed_mask = vec![0; current_mask_size];
            self.processed = vec![0; current_mask_size];
        }

        let stride_x = dim_y * dim_z;
        let max_dim = self.max_dim;
        let dims = [dim_x, dim_y, dim_z];

        for axis in 0..3usize {
            let u = (axis + 1) % 3;
            let v = (axis + 2) % 3;

            let axis_size = dims[axis];
            let u_size = dims[u];
            let v_size = dims[v];

            for dir_idx in 0..2usize {
                let dir: i32 = if dir_idx == 0 { -1 } else { 1 };
                let face_dir = axis * 2 + if dir > 0 { 0 } else { 1 };
                let dx: i32 = if axis == 0 { dir } else { 0 };
                let dy: i32 = if axis == 1 { dir } else { 0 };
                let dz: i32 = if axis == 2 { dir } else { 0 };
                let neighbor_max = dims[axis] as i32;

                let x_is_depth = axis == 0;
                let y_is_depth = axis == 1;
                let z_is_depth = axis == 2;
                let x_is_u_axis = !x_is_depth && u == 0;
                let y_is_u_axis = !y_is_depth && u == 1;
                let z_is_u_axis = !z_is_depth && u == 2;

                for d in 0..axis_size {
                    for iv in 0..v_size {
                        let row_offset = iv * max_dim;
                        for i in row_offset..row_offset + u_size {
                            self.packed_mask[i] = 0;
                        }
                    }

                    let mut has_faces = false;

                    for iu in 0..u_size {
                        for iv in 0..v_size {
                            let x = if x_is_depth { d } else if x_is_u_axis { iu } else { iv };
                            let y = if y_is_depth { d } else if y_is_u_axis { iu } else { iv };
                            let z = if z_is_depth { d } else if z_is_u_axis { iu } else { iv };

                            let block_value = voxel_data[x * stride_x + y * dim_z + z];
                            let block_type = block_value & 0x7F;
                            let block_visible = block_type != 0;
                            let block_is_selected = !selection_empty
                                && is_selection_set(
                                    selection_data,
                                    x,
                                    y,
                                    z,
                                    selection_dim_x,
                                    selection_dim_y,
                                    selection_dim_z,
                                );

                            if !block_visible && !block_is_selected {
                                continue;
                            }

                            let nx = x as i32 + dx;
                            let ny = y as i32 + dy;
                            let nz = z as i32 + dz;

                            let mask_idx = iv * max_dim + iu;

                            if block_is_selected && !block_visible {
                                let neighbor_coord = if x_is_depth {
                                    nx
                                } else if y_is_depth {
                                    ny
                                } else {
                                    nz
                                };
                                let neighbor_in_bounds = if dir > 0 {
                                    neighbor_coord < neighbor_max
                                } else {
                                    neighbor_coord >= 0
                                };
                                let neighbor_is_selected = neighbor_in_bounds
                                    && is_selection_set(
                                        selection_data,
                                        nx as usize,
                                        ny as usize,
                                        nz as usize,
                                        selection_dim_x,
                                        selection_dim_y,
                                        selection_dim_z,
                                    );

                                if !neighbor_is_selected {
                                    let selection_block_type = get_selection_value(
                                        selection_data,
                                        x,
                                        y,
                                        z,
                                        selection_dim_x,
                                        selection_dim_y,
                                        selection_dim_z,
                                    ) & 0x7F;
                                    let texture_index = block_atlas_mapping
                                        [(selection_block_type.max(1) - 1) as usize];

                                    let ao = calculate_ambient_occlusion(
                                        nx,
                                        ny,
                                        nz,
                                        face_dir,
                                        voxel_data,
                                        dim_x as i32,
                                        dim_y as i32,
                                        dim_z as i32,
                                        stride_x as i32,
                                    );

                                    self.packed_mask[mask_idx] = PACKED_PRESENT_BIT
                                        | (texture_index & 0x3FF)
                                        | ((ao as i32 & 0xFF) << 10)
                                        | (1 << 18);
                                    has_faces = true;
                                }
                            } else if block_visible {
                                let neighbor_coord = if x_is_depth {
                                    nx
                                } else if y_is_depth {
                                    ny
                                } else {
                                    nz
                                };
                                let neighbor_in_bounds = if dir > 0 {
                                    neighbor_coord < neighbor_max
                                } else {
                                    neighbor_coord >= 0
                                };
                                let neighbor_visible = neighbor_in_bounds
                                    && (voxel_data
                                        [(nx as usize) * stride_x + (ny as usize) * dim_z + (nz as usize)]
                                        & 0x7F)
                                        != 0;

                                if !neighbor_visible {
                                    let texture_index =
                                        block_atlas_mapping[(block_type - 1) as usize];

                                    let ao = calculate_ambient_occlusion(
                                        nx,
                                        ny,
                                        nz,
                                        face_dir,
                                        voxel_data,
                                        dim_x as i32,
                                        dim_y as i32,
                                        dim_z as i32,
                                        stride_x as i32,
                                    );

                                    let sel = if block_is_selected { 1i32 } else { 0i32 };
                                    self.packed_mask[mask_idx] = PACKED_PRESENT_BIT
                                        | (texture_index & 0x3FF)
                                        | ((ao as i32 & 0xFF) << 10)
                                        | (sel << 18);
                                    has_faces = true;
                                }
                            }
                        }
                    }

                    if has_faces {
                        self.generate_greedy_mesh(
                            u_size,
                            v_size,
                            d,
                            axis,
                            u,
                            v,
                            dir,
                            face_dir,
                            texture_width,
                            mesh_arrays,
                        );
                    }
                }
            }
        }
    }

    fn generate_greedy_mesh(
        &mut self,
        width: usize,
        height: usize,
        depth: usize,
        axis: usize,
        u: usize,
        v: usize,
        dir: i32,
        face_dir: usize,
        texture_width: i32,
        mesh_arrays: &mut MeshArrays,
    ) {
        let stride = self.max_dim;
        let normal = FACES[face_dir].normal;
        let face_offset: f32 = if dir > 0 { 1.0 } else { 0.0 };

        for iv in 0..height {
            let row_offset = iv * stride;
            for i in row_offset..row_offset + width {
                self.processed[i] = 0;
            }
        }

        let mut j = 0;
        while j < height {
            let j_offset = j * stride;
            let mut i = 0;
            while i < width {
                let ji = j_offset + i;
                if self.processed[ji] != 0 || (self.packed_mask[ji] & PACKED_PRESENT_BIT) == 0 {
                    i += 1;
                    continue;
                }

                let packed = self.packed_mask[ji];
                let texture_index = packed & 0x3FF;
                let ao_val = ((packed >> 10) & 0xFF) as u8;
                let is_selected = ((packed >> 18) & 0x1) as u8;
                let mut quad_width = 1usize;

                while i + quad_width < width {
                    let idx = j_offset + i + quad_width;
                    if self.processed[idx] != 0
                        || self.packed_mask[idx] != packed
                    {
                        break;
                    }
                    quad_width += 1;
                }

                let mut quad_height = 1usize;
                'outer: while j + quad_height < height {
                    let row_off = (j + quad_height) * stride;
                    for w in 0..quad_width {
                        let idx = row_off + i + w;
                        if self.processed[idx] != 0
                            || self.packed_mask[idx] != packed
                        {
                            break 'outer;
                        }
                    }
                    quad_height += 1;
                }

                let end_i = i + quad_width;
                let end_j = j + quad_height;
                for jj in j..end_j {
                    let row_off = jj * stride;
                    for ii in i..end_i {
                        self.processed[row_off + ii] = 1;
                    }
                }

                let x = if axis == 0 {
                    depth
                } else if u == 0 {
                    i
                } else if v == 0 {
                    j
                } else {
                    0
                };
                let y = if axis == 1 {
                    depth
                } else if u == 1 {
                    i
                } else if v == 1 {
                    j
                } else {
                    0
                };
                let z = if axis == 2 {
                    depth
                } else if u == 2 {
                    i
                } else if v == 2 {
                    j
                } else {
                    0
                };

                let texture_coords =
                    get_texture_coordinates(texture_index as i32, texture_width);

                let start_vertex_index = mesh_arrays.vertex_count as u32;

                for vi in 0..4u32 {
                    let actual_vi = if dir < 0 && (vi == 1 || vi == 3) {
                        if vi == 1 { 3 } else { 1 }
                    } else {
                        vi
                    };

                    let (vx, vy, vz): (f32, f32, f32);

                    if axis == 0 {
                        vx = x as f32 + face_offset;
                        match actual_vi {
                            0 => {
                                vy = y as f32;
                                vz = z as f32;
                            }
                            1 => {
                                vy = y as f32
                                    + if u == 1 { quad_width as f32 } else { 0.0 };
                                vz = z as f32
                                    + if u == 2 { quad_width as f32 } else { 0.0 };
                            }
                            2 => {
                                vy = y as f32
                                    + if u == 1 {
                                        quad_width as f32
                                    } else if v == 1 {
                                        quad_height as f32
                                    } else {
                                        0.0
                                    };
                                vz = z as f32
                                    + if u == 2 {
                                        quad_width as f32
                                    } else if v == 2 {
                                        quad_height as f32
                                    } else {
                                        0.0
                                    };
                            }
                            _ => {
                                vy = y as f32
                                    + if v == 1 { quad_height as f32 } else { 0.0 };
                                vz = z as f32
                                    + if v == 2 { quad_height as f32 } else { 0.0 };
                            }
                        }
                    } else if axis == 1 {
                        vy = y as f32 + face_offset;
                        match actual_vi {
                            0 => {
                                vx = x as f32;
                                vz = z as f32;
                            }
                            1 => {
                                vx = x as f32
                                    + if u == 0 { quad_width as f32 } else { 0.0 };
                                vz = z as f32
                                    + if u == 2 { quad_width as f32 } else { 0.0 };
                            }
                            2 => {
                                vx = x as f32
                                    + if u == 0 {
                                        quad_width as f32
                                    } else if v == 0 {
                                        quad_height as f32
                                    } else {
                                        0.0
                                    };
                                vz = z as f32
                                    + if u == 2 {
                                        quad_width as f32
                                    } else if v == 2 {
                                        quad_height as f32
                                    } else {
                                        0.0
                                    };
                            }
                            _ => {
                                vx = x as f32
                                    + if v == 0 { quad_height as f32 } else { 0.0 };
                                vz = z as f32
                                    + if v == 2 { quad_height as f32 } else { 0.0 };
                            }
                        }
                    } else {
                        vz = z as f32 + face_offset;
                        match actual_vi {
                            0 => {
                                vx = x as f32;
                                vy = y as f32;
                            }
                            1 => {
                                vx = x as f32
                                    + if u == 0 { quad_width as f32 } else { 0.0 };
                                vy = y as f32
                                    + if u == 1 { quad_width as f32 } else { 0.0 };
                            }
                            2 => {
                                vx = x as f32
                                    + if u == 0 {
                                        quad_width as f32
                                    } else if v == 0 {
                                        quad_height as f32
                                    } else {
                                        0.0
                                    };
                                vy = y as f32
                                    + if u == 1 {
                                        quad_width as f32
                                    } else if v == 1 {
                                        quad_height as f32
                                    } else {
                                        0.0
                                    };
                            }
                            _ => {
                                vx = x as f32
                                    + if v == 0 { quad_height as f32 } else { 0.0 };
                                vy = y as f32
                                    + if v == 1 { quad_height as f32 } else { 0.0 };
                            }
                        }
                    }

                    mesh_arrays.push_vertex(vx, vy, vz);
                    mesh_arrays.push_normal(normal[0], normal[1], normal[2]);
                    mesh_arrays.push_uv(
                        texture_coords[(vi * 2) as usize] as f32,
                        texture_coords[(vi * 2 + 1) as usize] as f32,
                    );

                    let packed_ao = ao_val;

                    let ao_corner_index =
                        if face_dir == 1 || face_dir == 2 || face_dir == 5 {
                            if vi == 1 { 3 } else if vi == 3 { 1 } else { vi }
                        } else {
                            vi
                        };

                    let occlusion_count =
                        (packed_ao >> (ao_corner_index * 2)) & 0x03;
                    let ao_factor = OCCLUSION_LEVELS[occlusion_count as usize];
                    mesh_arrays.push_ao(ao_factor);
                    mesh_arrays.push_is_selected(is_selected);
                    mesh_arrays.increment_vertex();
                }

                mesh_arrays.push_index(start_vertex_index);
                mesh_arrays.push_index(start_vertex_index + 1);
                mesh_arrays.push_index(start_vertex_index + 2);
                mesh_arrays.push_index(start_vertex_index);
                mesh_arrays.push_index(start_vertex_index + 2);
                mesh_arrays.push_index(start_vertex_index + 3);

                i += quad_width;
            }
            j += 1;
        }
    }
}

#[inline(always)]
fn is_selection_set(
    selection_data: &[u8],
    x: usize,
    y: usize,
    z: usize,
    sel_dim_x: usize,
    sel_dim_y: usize,
    sel_dim_z: usize,
) -> bool {
    if x >= sel_dim_x || y >= sel_dim_y || z >= sel_dim_z {
        return false;
    }
    selection_data[x * sel_dim_y * sel_dim_z + y * sel_dim_z + z] != 0
}

#[inline(always)]
fn get_selection_value(
    selection_data: &[u8],
    x: usize,
    y: usize,
    z: usize,
    sel_dim_x: usize,
    sel_dim_y: usize,
    sel_dim_z: usize,
) -> u8 {
    if x >= sel_dim_x || y >= sel_dim_y || z >= sel_dim_z {
        return 0;
    }
    selection_data[x * sel_dim_y * sel_dim_z + y * sel_dim_z + z]
}
